text/1025-branches.rst

```diff
@@ -0,0 +1,181 @@
+===========================
+
+Abstract
+========
+
+We propose to reframe the current concept of ``database`` in EdgeDB as
+``branch``. This avoids some of the confusion in current terminology and
+provides a clearer guideline for using EdgeDB in development environment.
+Functionally a ``branch`` will be the same as the current ``database``, but
+with some additional options and tooling.
+
+
+Motivation
+==========
+
+The current nomenclature of ``database`` and the corresponding commands are
+confusing to many users. A running EdgeDB server represents an *instance*
+while each instance can have multiple *databases* in it. Since the databases
+all run by the same server they all have to use the same underlying EdgeDB
+version. These qualities make them impractical for usage across multiple
+projects as the projects would have to share resources and perform EdgeDB
+upgrades together. We specifically recommend to have a dedicated instance for
+each project. This leaves databases as a feature without a clear usage.
+
+Instead we want to enhance the developer experience while using multiple
+source control branches. These multiple databases could map well onto
+corresponding source branches, representing slightly different version of the
+data and schema. Thus there is a good reason to rename the concept of
+*database* to *branch* and provide additional branch-handling tools. With this
+change it is also now no longer necessary to have a semantic distinction
+between the terms *instance* and *database* which may be natural for some
+users to think of as interchangeable.
+
+
+Branch DDL Commands
+===================
+
+There are several new DDL commands introduced by this proposal:
+
+1) ``create empty branch <newbranch>``
+
+   The most basic command creates a new branch with an empty schema
+   (exactly what ``create database`` command currently does).
+
+2) ``create branch <newbranch> [ from <oldbranch> ]``
+
+   This command creates a new branch and copies the schema of an existing
+   branch to it. Only the schema is copied, the data is still empty and needs
+   to be populated separately. This parallels source control branches that
+   diverge in terms of the schema. It is recommended to setup a new database
+   branch whenever working on a code branch that alters the schema in some
+   way. This will help reduce friction when switching between code branches by
+   having database states that correspond to the schema and avoiding
+   accidental mismatches in migrations.
+
+   If the ``from <oldbranch>`` part is omitted, the new branch is created
+   using the branch from the current connection.
+
+3) ``drop branch <oldbranch>``
+
+   Removes an existing branch from the instance.
+
+
+Project and CLI Integration
+---------------------------
+
+Currently the project configuration file ``edgedb.toml`` does not have to
+specify the default *database* it connects to. With this proposal specifying
+the branch is mandatory.
```

Comment:
This might want to expand the reasoning for this:

> Intended usage is to create a new database branch when work on a new VCS branch is started. When switching VCS branches, `edgedb.toml` would be changed as well, changing the used database branch.

Reply:
The RFC hasn't been updated yet, but we decided on the call yesterday that storing the current branch in `edgedb.toml` is probably a mistake, because it will be too easy to accidentally merge the config for a feature branch back into main.

What we were thinking of doing instead in storing the db branch name in `.edgedb.local.toml` or some such, which should be in `.gitignore`. The CLI could offer to put it in `.gitignore` for you, and it can also offer to install a `post-checkout` hook script that automatically updates `.edgedb.local.toml` when changing branches.